lib/handler/mruby/middleware.c

```diff
@@ -0,0 +1,961 @@
+        goto Failed;
+    }
+
+    if (!is_reprocess) {
+        /* ensure that SCRIPT_NAME is not modified */
+        h2o_iovec_t confpath = ctx->handler->pathconf->path;
+        size_t confpath_len_wo_slash = confpath.base[confpath.len - 1] == '/' ? confpath.len - 1 : confpath.len;
+        if (!(RSTRING_LEN(script_name) == confpath_len_wo_slash && memcmp(RSTRING_PTR(script_name), confpath.base, confpath_len_wo_slash) == 0)) {
+            mrb->exc = mrb_obj_ptr(mrb_exc_new_str_lit(mrb, E_RUNTIME_ERROR, "can't modify `SCRIPT_NAME` with `H2O.next`. Is `H2O.reprocess` what you want?"));
+            goto Failed;
+        }
+    }
+
+#define STR_TO_IOVEC(val) h2o_iovec_init(RSTRING_PTR(val), RSTRING_LEN(val))
+
+    /* construct url and parse */
+    h2o_iovec_t *url_comps = alloca(sizeof(*url_comps) * 9);
+    int num_comps = 0;
+    url_comps[num_comps++] = STR_TO_IOVEC(scheme);
+    url_comps[num_comps++] = h2o_iovec_init(H2O_STRLIT("://"));
+    if (!mrb_nil_p(http_host)) {
+        url_comps[num_comps++] = STR_TO_IOVEC(http_host);
+    } else {
+        url_comps[num_comps++] = STR_TO_IOVEC(server_name);
+        url_comps[num_comps++] = h2o_iovec_init(H2O_STRLIT(":"));
+        url_comps[num_comps++] = STR_TO_IOVEC(server_port);
+    }
+    url_comps[num_comps++] = STR_TO_IOVEC(script_name);
+    url_comps[num_comps++] = STR_TO_IOVEC(path_info);
+    if (RSTRING_LEN(query_string) != 0) {
+        url_comps[num_comps++] = h2o_iovec_init(H2O_STRLIT("?"));
+        url_comps[num_comps++] = STR_TO_IOVEC(query_string);
+    }
+    h2o_iovec_t url_str = h2o_concat_list(&super->pool, url_comps, num_comps);
+    h2o_url_t url_parsed;
+    if (h2o_url_parse(url_str.base, url_str.len, &url_parsed) != 0) {
+        /* TODO is there any other way to show better error message? */
+        mrb->exc = mrb_obj_ptr(mrb_exc_new_str_lit(mrb, E_ARGUMENT_ERROR, "env variable contains invalid values"));
+        goto Failed;
+    }
+
+    /* setup req and conn using retrieved values */
+    super->input.scheme = url_parsed.scheme;
+    super->input.method = h2o_strdup(&super->pool, RSTRING_PTR(method), RSTRING_LEN(method));
+    super->input.authority = h2o_strdup(&super->pool, url_parsed.authority.base, url_parsed.authority.len);
+    super->input.path = h2o_strdup(&super->pool, url_parsed.path.base, url_parsed.path.len);
+    h2o_hostconf_t *hostconf = h2o_req_setup(super);
+    super->hostconf = hostconf;
+    super->pathconf = ctx->handler->pathconf;
+    super->handler = &ctx->handler->super;
+    super->version = h2o_parse_protocol_version(STR_TO_IOVEC(server_protocol));
+    if (super->version == -1)
+        super->version = 0x101;
+
+    if (!(mrb_nil_p(server_addr) || mrb_nil_p(server_port))) {
+        subreq->conn.server.host = h2o_strdup(&super->pool, RSTRING_PTR(server_addr), RSTRING_LEN(server_addr));
+        subreq->conn.server.port = h2o_strdup(&super->pool, RSTRING_PTR(server_port), RSTRING_LEN(server_port));
```

Comment:
Is there a guarantee that both of the variables will be non-nil if either one is?

If not, we have a chance of calling `RSTRING_PTR`, `RSTRING_LEN` against nil. Is that fine?